crates/forge/src/lib.rs

```diff
@@ -462,35 +488,101 @@ fn run_with_fuzzing(
     };
     let mut fuzzer = RandomFuzzer::create(fuzzer_seed, fuzzer_runs, &args)?;
 
-    let mut results = vec![];
+    let mut tasks = FuturesUnordered::new();
+
+    let (send, mut recv) = channel(1);
 
     for _ in 1..=fuzzer_runs {
         let args = fuzzer.next_args();
+        let case = case.clone();
+        let case_copy = case.clone();
+        let runner = runner.clone();
+        let args = args.clone();
+        let cancellation_fuzzing_token = token.clone();
+        let runner_config = runner_config.clone();
+        let send = send.clone();
+
+        tasks.push(task::spawn(async move {
+            tokio::select! {
+                () = runner_config.error_cancellation_token.cancelled() => {
+                    //Stop executing all tests because
+                    //one of a test returns Err
+                    Ok(TestCaseSummary::None {  })
+                },
+                () = runner_config.exit_first_cancellation_token.cancelled() => {
+                    //Stop executing all tests because flag --exit-first'
+                    //has been set and one test FAIL
+                    Ok(TestCaseSummary::skipped(&case_copy))
+                },
+                () = cancellation_fuzzing_token.cancelled() => {
+                    //Stop executing all single fuzzing tests
+                    //because one of fuzzing test has been FAIL
+                    Ok(TestCaseSummary::skipped(&case_copy))
+                },
+               result = blocking_run_from_test(
+                    runner,
+                    case,
+                    runner_config.clone(),
+                    args.clone(),
+                    Some(send),
+                ) => {
+                    match result {
+                        Ok(result) => {
+                            if let TestCaseSummary::Failed { .. } = &result {
+                                if runner_config.exit_first {
+                                    runner_config.exit_first_cancellation_token.cancel();
+                                } else {
+                                    cancellation_fuzzing_token.cancel();
+                                }
+                            }
+                            Ok(result)
+                        }
+                        Err(e) => {
+                            runner_config.error_cancellation_token.cancel();
+                            Err(e)
+                        }
+                    }
+                },
+            }
+        }));
+    }
 
-        let result = run_from_test_case(
-            package_root,
-            runner,
-            case,
-            runner_config.fork_targets.as_ref(),
-            &runner_params.contracts,
-            &runner_params.predeployed_contracts,
-            args.clone(),
-            &runner_params.environment_variables,
-        )?;
+    let mut results = vec![];
+
+    drop(send);
```

Comment:
This could use some explanation/linking to tokio docs, so someone doesn't accidentally remove this line in the future.